File: backend/app/schemas/projeto_vida_mensal.py
"""
Schemas — Projeto de Vida Mensal
=================================
"""

from datetime import datetime
from typing import Any, List, Optional
from uuid import UUID

from pydantic import BaseModel, Field, field_validator


# ── Sub-schemas: itens de lista ─────────────────────────────────────────────


class EventoItem(BaseModel):
    """Compromisso/encontro com data, horário e local. Usado em comunidade."""
    data: Optional[str] = Field(None, max_length=20)        # "2026-04-04"
    horario: Optional[str] = Field(None, max_length=20)     # "19:00"
    local: Optional[str] = Field(None, max_length=500)
    observacoes: Optional[str] = Field(None, max_length=2000)


class OutroItemComunidade(BaseModel):
    """Item genérico de compromisso comunitário."""
    titulo: Optional[str] = Field(None, max_length=200)
    descricao: Optional[str] = Field(None, max_length=2000)
    local: Optional[str] = Field(None, max_length=500)
    data: Optional[str] = Field(None, max_length=20)
    horario: Optional[str] = Field(None, max_length=20)


class CuidadoEventoItem(BaseModel):
    """Consulta / exame / descanso com data, horário e local."""
    data: Optional[str] = Field(None, max_length=20)
    horario: Optional[str] = Field(None, max_length=20)
    local: Optional[str] = Field(None, max_length=500)
    descricao: Optional[str] = Field(None, max_length=2000)


class OutroItemCuidado(BaseModel):
    """Item genérico de cuidado pessoal."""
    titulo: Optional[str] = Field(None, max_length=200)
    descricao: Optional[str] = Field(None, max_length=2000)
    local: Optional[str] = Field(None, max_length=500)
    data: Optional[str] = Field(None, max_length=20)
    horario: Optional[str] = Field(None, max_length=20)


# ── Sub-schemas de seção ────────────────────────────────────────────────────


class ComunidadeData(BaseModel):
    partilha_acompanhador: List[EventoItem] = []
    encontro_familia: List[EventoItem] = []
    dias_grupo: List[EventoItem] = []
    outros: List[OutroItemComunidade] = []


class CuidadoData(BaseModel):
    consultas: List[CuidadoEventoItem] = []
    exames: List[CuidadoEventoItem] = []
    descanso: List[CuidadoEventoItem] = []
    outros: List[OutroItemCuidado] = []


class CompromissoIn(BaseModel):
    semana: str = Field(..., pattern=r"^s[1-5]$")
    titulo: Optional[str] = Field(None, max_length=200)
    dia: Optional[str] = Field(None, max_length=50)
    horario: Optional[str] = Field(None, max_length=50)
    obs: Optional[str] = Field(None, max_length=1000)
    ordem: int = 0


class CompromissoOut(BaseModel):
    id: UUID
    semana: str
    titulo: Optional[str] = None
    dia: Optional[str] = None
    horario: Optional[str] = None
    obs: Optional[str] = None
    ordem: int

    model_config = {"from_attributes": True}


class PraticaIn(BaseModel):
    dia_semana: str = Field(..., pattern=r"^(seg|ter|qua|qui|sex|sab|dom)$")
    tipo: str = Field(..., max_length=100)
    horario: Optional[str] = Field(None, max_length=50)
    duracao: Optional[str] = Field(None, max_length=50)
    obs: Optional[str] = Field(None, max_length=1000)
    ordem: int = 0


class PraticaOut(BaseModel):
    id: UUID
    dia_semana: str
    tipo: str
    horario: Optional[str] = None
    duracao: Optional[str] = None
    obs: Optional[str] = None
    ordem: int

    model_config = {"from_attributes": True}


class RevisaoUpsert(BaseModel):
    """4 perguntas da revisão mensal v2."""
    pratica_melhorar: Optional[str] = Field(None, max_length=5000)
    taticas_vigilancia: Optional[str] = Field(None, max_length=5000)
    rotina_evangelizacao: Optional[str] = Field(None, max_length=5000)
    outra_area_atencao: Optional[str] = Field(None, max_length=5000)


class RevisaoOut(BaseModel):
    pratica_melhorar: Optional[str] = None
    taticas_vigilancia: Optional[str] = None
    rotina_evangelizacao: Optional[str] = None
    outra_area_atencao: Optional[str] = None

    model_config = {"from_attributes": True}


# ── Áreas mensais ───────────────────────────────────────────────────────────

TIPOS_AREA_VALIDOS = frozenset({
    "FAMILIA_VOCACIONAL",
    "MINISTERIO_BOM_PASTOR",
    "GRUPO_FORMATIVO",
    "SAUDE_LAZER",
    "FAMILIA_ORIGEM",
})


class CompromissoAreaItem(BaseModel):
    descricao: Optional[str] = Field(None, max_length=500)
    data: Optional[str] = Field(None, max_length=20)
    horario: Optional[str] = Field(None, max_length=20)
    local: Optional[str] = Field(None, max_length=300)
    obs: Optional[str] = Field(None, max_length=1000)


class AreaMensalIn(BaseModel):
    tipo_area: str
    objetivo: Optional[str] = Field(None, max_length=3000)
    compromissos: Optional[List[CompromissoAreaItem]] = []
    observacoes: Optional[str] = Field(None, max_length=3000)

    @field_validator("tipo_area")
    @classmethod
    def validar_tipo_area(cls, v: str) -> str:
        if v not in TIPOS_AREA_VALIDOS:
            raise ValueError(
                f"tipo_area inválido: '{v}'. Valores aceitos: {sorted(TIPOS_AREA_VALIDOS)}"
            )
        return v


class AreaMensalOut(BaseModel):
    id: UUID
    tipo_area: str
    objetivo: Optional[str] = None
    compromissos: Optional[List[CompromissoAreaItem]] = []
    observacoes: Optional[str] = None

    model_config = {"from_attributes": True}


# ── Top-level schemas ──────────────────────────────────────────────────────


class ProjetoVidaMensalCreate(BaseModel):
    mes: int = Field(..., ge=1, le=12)
    ano: int = Field(..., ge=2024, le=2100)
    pin: Optional[str] = Field(None, min_length=4, max_length=4, pattern=r"^\d{4}$")
    intencao: Optional[str] = Field(None, max_length=2000)
    reflexao_evangelizacao: Optional[str] = Field(None, max_length=3000)


class ProjetoVidaMensalUpdate(BaseModel):
    observacoes_mes: Optional[str] = Field(None, max_length=3000)
    concluido: Optional[bool] = None
    intencao: Optional[str] = Field(None, max_length=2000)
    reflexao_evangelizacao: Optional[str] = Field(None, max_length=3000)
    areas: Optional[List[AreaMensalIn]] = None
    comunidade: Optional[ComunidadeData] = None
    cuidado: Optional[CuidadoData] = None
    compromissos: Optional[List[CompromissoIn]] = None
    praticas: Optional[List[PraticaIn]] = None


class ProjetoVidaMensalSummary(BaseModel):
    id: UUID
    mes: int
    ano: int
    concluido: bool
    has_pin: bool
    created_at: datetime


class ProjetoVidaMensalFull(BaseModel):
    id: UUID
    mes: int
    ano: int
    has_pin: bool
    concluido: bool
    observacoes_mes: Optional[str] = None
    reflexao_evangelizacao: Optional[str] = None
    has_new_structure: bool = False
    areas: List[AreaMensalOut] = []
    intencao: Optional[str] = None
    comunidade: Optional[ComunidadeData] = None
    cuidado: Optional[CuidadoData] = None
    compromissos: List[CompromissoOut] = []
    praticas: List[PraticaOut] = []
    revisao: Optional[RevisaoOut] = None
    created_at: datetime
    updated_at: datetime


class PinVerifyRequest(BaseModel):
    pin: str = Field(..., min_length=4, max_length=4, pattern=r"^\d{4}$")


class PinVerifyResponse(BaseModel):
    valid: bool


# ── Exame de Consciência ────────────────────────────────────────────────────


class ExameUpsert(BaseModel):
    """6 campos de reflexão — todos opcionais (exame nunca é obrigatório)."""
    gracas_recebidas: Optional[str] = Field(None, max_length=5000)
    infidelidades: Optional[str] = Field(None, max_length=5000)
    dificuldades_espirituais: Optional[str] = Field(None, max_length=5000)
    jesus_abandonado: Optional[str] = Field(None, max_length=5000)
    onde_deixei_de_responder: Optional[str] = Field(None, max_length=5000)
    proposito_conversao: Optional[str] = Field(None, max_length=5000)


class ExameOut(BaseModel):
    id: UUID
    gracas_recebidas: Optional[str] = None
    infidelidades: Optional[str] = None
    dificuldades_espirituais: Optional[str] = None
    jesus_abandonado: Optional[str] = None
    onde_deixei_de_responder: Optional[str] = None
    proposito_conversao: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ── Intercessão ─────────────────────────────────────────────────────────────


class IntercessaoUpsert(BaseModel):
    intencoes_pessoais: Optional[str] = Field(None, max_length=5000)
    intencoes_comunitarias: Optional[str] = Field(None, max_length=5000)
    oferecimento: Optional[str] = Field(None, max_length=3000)


class IntercessaoOut(BaseModel):
    id: UUID
    intencoes_pessoais: Optional[str] = None
    intencoes_comunitarias: Optional[str] = None
    oferecimento: Optional[str] = None
    created_at: datetime
    updated_at: datetime


# ── Projeto Semanal ─────────────────────────────────────────────────────────


class EvangelizacaoMomentoItem(BaseModel):
    """Momento relacional planejado — apenas descrição, sem métricas."""
    descricao: str = Field(..., max_length=500)


class PlanoDiarioItem(BaseModel):
    """Planejamento de um dia — todos os campos opcionais."""
    proposito: Optional[str] = Field(None, max_length=1000)
    missa: Optional[bool] = None
    horario_missa: Optional[str] = Field(None, max_length=20)
    oracao_manha: Optional[str] = Field(None, max_length=500)
    lectio: Optional[str] = Field(None, max_length=500)
    terco: Optional[bool] = None
    leitura_espiritual: Optional[str] = Field(None, max_length=500)
    evangelizacao: Optional[str] = Field(None, max_length=500)
    compromissos: Optional[List[str]] = None


class ProjetoVidaSemanasCreate(BaseModel):
    numero_semana: int = Field(..., ge=1, le=5)
    dever_estado: Optional[Any] = None
    vida_interior: Optional[Any] = None
    evangelizacao_disposicao: Optional[str] = Field(None, max_length=3000)
    evangelizacao_momentos: Optional[List[EvangelizacaoMomentoItem]] = None
    plano_diario: Optional[dict[str, PlanoDiarioItem]] = None
    observacoes: Optional[str] = Field(None, max_length=2000)


class ProjetoVidaSemanasUpdate(BaseModel):
    dever_estado: Optional[Any] = None
    vida_interior: Optional[Any] = None
    evangelizacao_disposicao: Optional[str] = Field(None, max_length=3000)
    evangelizacao_momentos: Optional[List[EvangelizacaoMomentoItem]] = None
    plano_diario: Optional[dict[str, PlanoDiarioItem]] = None
    observacoes: Optional[str] = Field(None, max_length=2000)


class ProjetoVidaSemanasOut(BaseModel):
    id: UUID
    numero_semana: int
    dever_estado: Optional[Any] = None
    vida_interior: Optional[Any] = None
    evangelizacao_disposicao: Optional[str] = None
    evangelizacao_momentos: Optional[List[EvangelizacaoMomentoItem]] = None
    plano_diario: Optional[dict] = None
    observacoes: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ProjetoVidaSemanasSummary(BaseModel):
    """Sumário sem plano_diario para listagem."""
    id: UUID
    numero_semana: int
    created_at: datetime

    model_config = {"from_attributes": True}

    model_config = {"from_attributes": True}

File: lumen_mobile/src/services/projetoVidaMensal.ts
// lumen_mobile/src/services/projetoVidaMensal.ts
import api from '@/services/api';

// ── Constantes ─────────────────────────────────────────────────────────────

export const MESES = [
  'Janeiro','Fevereiro','Março','Abril','Maio','Junho',
  'Julho','Agosto','Setembro','Outubro','Novembro','Dezembro',
];

export const SEMANAS = ['s1','s2','s3','s4','s5'] as const;
export const SEMANA_LABELS: Record<string, string> = {
  s1: 'Semana 1', s2: 'Semana 2', s3: 'Semana 3', s4: 'Semana 4', s5: 'Semana 5',
};

export const DIAS = ['seg','ter','qua','qui','sex','sab','dom'] as const;
export const DIA_LABELS: Record<string, string> = {
  seg: 'Segunda', ter: 'Terça', qua: 'Quarta', qui: 'Quinta',
  sex: 'Sexta',   sab: 'Sábado', dom: 'Domingo',
};

export const TIPOS_PRATICA = [
  'Santa Missa',
  'Adoração',
  'Terço',
  'Leitura espiritual',
  'Liturgia das Horas',
  'Meditação',
  'Momento de Evangelização Ser Feliz',
  'Outro',
];

// ── Tipos de itens estruturados ────────────────────────────────────────────

export interface EventoItem {
  data: string | null;
  horario: string | null;
  local: string | null;
  observacoes: string | null;
}

export interface OutroItemComunidade {
  titulo: string | null;
  descricao: string | null;
  local: string | null;
  data: string | null;
  horario: string | null;
}

export interface CuidadoEventoItem {
  data: string | null;
  horario: string | null;
  local: string | null;
  descricao: string | null;
}

export interface OutroItemCuidado {
  titulo: string | null;
  descricao: string | null;
  local: string | null;
  data: string | null;
  horario: string | null;
}

export interface EvangelizacaoAcaoItem {
  descricao: string | null;
  como: string | null;
  duracao_min: number | null;
}

// ── Types principais ───────────────────────────────────────────────────────

export interface CompromissoOut {
  id: string;
  semana: string;
  titulo: string | null;
  dia: string | null;
  horario: string | null;
  obs: string | null;
  ordem: number;
}

export interface PraticaOut {
  id: string;
  dia_semana: string;
  tipo: string;
  horario: string | null;
  duracao: string | null;
  obs: string | null;
  ordem: number;
}

export interface ComunidadeData {
  partilha_acompanhador: EventoItem[];
  encontro_familia: EventoItem[];
  dias_grupo: EventoItem[];
  outros: OutroItemComunidade[];
}

export interface CuidadoData {
  consultas: CuidadoEventoItem[];
  exames: CuidadoEventoItem[];
  descanso: CuidadoEventoItem[];
  outros: OutroItemCuidado[];
}

export interface RevisaoOut {
  pratica_melhorar: string | null;
  taticas_vigilancia: string | null;
  rotina_evangelizacao: string | null;
  outra_area_atencao: string | null;
}

export interface ProjetoVidaMensalFull {
  id: string;
  mes: number;
  ano: number;
  has_pin: boolean;
  concluido: boolean;
  observacoes_mes: string | null;
  comunidade: ComunidadeData | null;
  cuidado: CuidadoData | null;
  compromissos: CompromissoOut[];
  praticas: PraticaOut[];
  revisao: RevisaoOut | null;
  has_new_structure: boolean;
  areas: AreaMensalOut[];
  reflexao_evangelizacao: string | null;
  evangelizacao_acoes?: EvangelizacaoAcaoItem[] | null;
  created_at: string;
  updated_at: string;
}

export interface ProjetoVidaMensalSummary {
  id: string;
  mes: number;
  ano: number;
  concluido: boolean;
  has_pin: boolean;
  created_at: string;
}

// ── Input types ────────────────────────────────────────────────────────────

export interface CompromissoIn {
  semana: string;
  titulo: string;
  dia: string;
  horario: string;
  obs: string;
  ordem: number;
}

export interface PraticaIn {
  dia_semana: string;
  tipo: string;
  horario: string;
  duracao: string;
  obs: string;
  ordem: number;
}

export interface CreateProjetoInput {
  mes: number;
  ano: number;
  pin?: string | null;
  intencao?: string | null;
}

export interface CompromissoAreaItem {
  descricao: string | null;
  data: string | null;
  horario: string | null;
  local: string | null;
  obs: string | null;
}

export interface AreaMensalIn {
  tipo_area: string;
  objetivo: string | null;
  compromissos: CompromissoAreaItem[];
  observacoes: string | null;
}

export interface AreaMensalOut extends AreaMensalIn {
  id: string;
}

export interface UpdateProjetoInput {
  observacoes_mes?: string | null;
  concluido?: boolean | null;
  comunidade?: ComunidadeData | null;
  cuidado?: CuidadoData | null;
  compromissos?: CompromissoIn[] | null;
  praticas?: PraticaIn[] | null;
  areas?: AreaMensalIn[] | null;
  reflexao_evangelizacao?: string | null;
  evangelizacao_acoes?: EvangelizacaoAcaoItem[] | null;
}

export interface RevisaoInput {
  pratica_melhorar?: string | null;
  taticas_vigilancia?: string | null;
  rotina_evangelizacao?: string | null;
  outra_area_atencao?: string | null;
}

export interface ProjetoVidaSemanasCreate {
  numero_semana: number;
  dever_estado?: object | null;
  vida_interior?: object | null;
  evangelizacao_disposicao?: string | null;
  evangelizacao_momentos?: Array<{ descricao: string }>;
  plano_diario?: Record<string, object>;
  observacoes?: string | null;
}

export interface ProjetoVidaSemanasOut {
  id: string;
  numero_semana: number;
  dever_estado?: object | null;
  vida_interior?: object | null;
  evangelizacao_disposicao?: string | null;
  evangelizacao_momentos?: Array<{ descricao: string }>;
  plano_diario?: Record<string, object> | null;
  observacoes?: string | null;
  created_at: string;
  updated_at: string;
}

export interface ProjetoVidaSemanasSummary {
  id: string;
  numero_semana: number;
  created_at: string;
}

export interface ContextoVocacionalOut {
  vocational_reality_code: string | null;
  life_state_code: string | null;
  perfil_incompleto: boolean;
  nome: string;
}

export interface ExameOut {
  id: string;
  gracas_recebidas: string | null;
  infidelidades: string | null;
  dificuldades_espirituais: string | null;
  jesus_abandonado: string | null;
  onde_deixei_de_responder: string | null;
  proposito_conversao: string | null;
  created_at: string;
  updated_at: string;
}

export interface ExameUpsert {
  gracas_recebidas?: string | null;
  infidelidades?: string | null;
  dificuldades_espirituais?: string | null;
  jesus_abandonado?: string | null;
  onde_deixei_de_responder?: string | null;
  proposito_conversao?: string | null;
}

export interface IntercessaoOut {
  id: string;
  intencoes_pessoais: string | null;
  intencoes_comunitarias: string | null;
  oferecimento: string | null;
  created_at: string;
  updated_at: string;
}

export interface IntercessaoUpsert {
  intencoes_pessoais?: string | null;
  intencoes_comunitarias?: string | null;
  oferecimento?: string | null;
}

export interface PlanoDiarioItem {
  proposito?: string | null;
  missa?: boolean | null;
  horario_missa?: string | null;
  oracao_manha?: string | null;
  lectio?: string | null;
  terco?: boolean | null;
  leitura_espiritual?: string | null;
  evangelizacao?: string | null;
  compromissos?: string[] | null;
}

// ── API ────────────────────────────────────────────────────────────────────

export const projetoVidaMensalApi = {
  getAtual: (mes?: number, ano?: number) => {
    const params = mes && ano ? `?mes=${mes}&ano=${ano}` : '';
    return api.get<ProjetoVidaMensalFull | null>(`/projeto-vida-mensal/atual${params}`);
  },

  getHistorico: () =>
    api.get<ProjetoVidaMensalSummary[]>('/projeto-vida-mensal/historico'),

  criar: (data: CreateProjetoInput) =>
    api.post<ProjetoVidaMensalFull>('/projeto-vida-mensal/', data as Record<string, unknown>),

  get: (id: string) =>
    api.get<ProjetoVidaMensalFull>(`/projeto-vida-mensal/${id}`),

  update: (id: string, data: UpdateProjetoInput) =>
    api.put<ProjetoVidaMensalFull>(`/projeto-vida-mensal/${id}`, data as Record<string, unknown>),

  upsertRevisao: (id: string, data: RevisaoInput) =>
    api.put<ProjetoVidaMensalFull>(`/projeto-vida-mensal/${id}/revisao`, data as Record<string, unknown>),

  verificarPin: (id: string, pin: string) =>
    api.post<{ valid: boolean }>(`/projeto-vida-mensal/${id}/pin/verificar`, { pin }),

  getContextoVocacional: () =>
    api.get<ContextoVocacionalOut>('/projeto-vida-mensal/contexto-vocacional'),

  getExame: (id: string) =>
    api.get<ExameOut | null>(`/projeto-vida-mensal/${id}/exame`),

  upsertExame: (id: string, data: ExameUpsert) =>
    api.put<ExameOut>(`/projeto-vida-mensal/${id}/exame`, data as Record<string, unknown>),

  upsertIntercessao: (id: string, data: IntercessaoUpsert) =>
    api.put<IntercessaoOut>(`/projeto-vida-mensal/${id}/intercessao`, data as Record<string, unknown>),

  listSemanas: (projetoId: string) =>
    api.get<ProjetoVidaSemanasSummary[]>(`/projeto-vida-mensal/${projetoId}/semanal`),

  createSemanal: (projetoId: string, data: ProjetoVidaSemanasCreate) =>
    api.post<ProjetoVidaSemanasOut>(`/projeto-vida-mensal/${projetoId}/semanal`, data as Record<string, unknown>),

  getSemanal: (id: string) =>
    api.get<ProjetoVidaSemanasOut>(`/projeto-vida-semanal/${id}`),

  updateSemanal: (id: string, data: Partial<ProjetoVidaSemanasCreate>) =>
    api.put<ProjetoVidaSemanasOut>(`/projeto-vida-semanal/${id}`, data as Record<string, unknown>),
};

export default projetoVidaMensalApi;
